src/main/java/com/mindee/input/URLInputSource.java

```diff
@@ -0,0 +1,231 @@
+    if (responseCode != HttpURLConnection.HTTP_OK) {
+      throw new IOException("Failed to fetch file: " + responseCode);
+    }
+
+    try (InputStream in = connection.getInputStream()) {
+      saveTempFile(in);
+    }
+  }
+
+  protected HttpURLConnection createConnection(String urlString) throws IOException {
+    HttpURLConnection connection = (HttpURLConnection) new URL(urlString).openConnection();
+    connection.setInstanceFollowRedirects(true);
+
+    if (username != null && password != null) {
+      String encodedCredentials =
+          Base64.getEncoder().encodeToString((username + ":" + password).getBytes());
+      connection.setRequestProperty("Authorization", "Basic " + encodedCredentials);
+    }
+    if (token != null) {
+      connection.setRequestProperty("Authorization", "Bearer " + token);
+    }
+    connection.setRequestMethod("GET");
+    return connection;
+  }
+
+  private HttpURLConnection handleRedirects(HttpURLConnection connection) throws IOException {
+    int status = connection.getResponseCode();
+    if (status == HttpURLConnection.HTTP_MOVED_TEMP
+        || status == HttpURLConnection.HTTP_MOVED_PERM
+        || status == HttpURLConnection.HTTP_SEE_OTHER
+        || status == 307
+        || status == 308
+    ) {
+      String newUrl = connection.getHeaderField("Location");
+      connection.disconnect();
+
+      HttpURLConnection newConnection = createConnection(newUrl);
+      return handleRedirects(newConnection);  // Recursive call to handle multiple redirects
+    }
+    return connection;
+  }
+
+  private void saveTempFile(InputStream in) throws IOException {
+    String filename = generateDefaultFilename();
+
+    Path tempFile = Files.createTempFile(filename, ".tmp");
+    localFilename = tempFile.toString();
+
+    try (InputStream inputStream = in;
+         OutputStream outputStream = Files.newOutputStream(tempFile)) {
+      byte[] buffer = new byte[4096];
+      int bytesRead;
+      while ((bytesRead = inputStream.read(buffer)) != -1) {
+        outputStream.write(buffer, 0, bytesRead);
+      }
+    }
+  }
+
+
+  private void saveFile(InputStream in, String filepath) throws IOException {
+    File outputFile = new File(filepath);
+
+    try (FileOutputStream out = new FileOutputStream(outputFile)) {
+      byte[] buffer = new byte[4096];
+      int bytesRead;
+      while ((bytesRead = in.read(buffer)) != -1) {
+        out.write(buffer, 0, bytesRead);
+      }
+    }
+  }
+
+  /**
+   * Create a LocalInputSource instance from this object.
+   *
+   * @return An instance of a {@link LocalInputSource}.
+   * @throws IOException Throws if the file can't be accessed.
+   */
+  public LocalInputSource toLocalInputSource() throws IOException {
+    File file = new File(localFilename);
+    return new LocalInputSource(file);
+  }
+
+  private String generateDefaultFilename() {
+    return "mindee_temp_"
+        + LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss"));
+  }
```

Comment:
Multiple files can be processed during the same second (especially if multi-threading).

some ideas:
* `mindee_temp_` + 8 random alphanumeric characters + `_` + yyyyMMdd_HHmmss
* UUID
* sha1 of the URL

Reply:
That's just a prefix, there won't be duplicates. I'll clarify the var names.